qtconsole/history_console_widget.py

```diff
@@ -76,31 +76,31 @@ def _up_pressed(self, shift_modifier):
                 return False
 
             # Set a search prefix based on the cursor position.
-            col = self._get_input_buffer_cursor_column()
+            pos = self._get_input_buffer_cursor_position_in_block()
             input_buffer = self.input_buffer
             # use the *shortest* of the cursor column and the history prefix
             # to determine if the prefix has changed
-            n = min(col, len(self._history_prefix))
-            
+            n = min(pos, len(self._history_prefix))
+
             # prefix changed, restart search from the beginning
             if (self._history_prefix[:n] != input_buffer[:n]):
                 self._history_index = len(self._history)
-            
+
             # the only time we shouldn't set the history prefix
             # to the line up to the cursor is if we are already
             # in a simple scroll (no prefix),
             # and the cursor is at the end of the first line
-            
+
             # check if we are at the end of the first line
             c = self._get_cursor()
             current_pos = c.position()
-            c.movePosition(QtGui.QTextCursor.EndOfLine)
+            c.movePosition(QtGui.QTextCursor.EndOfBlock)
```

Comment:
also change line 116/117 to: `cursor.movePosition(QtGui.QTextCuresor.EndOfBlock)`